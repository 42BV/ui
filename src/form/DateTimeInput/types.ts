// See https://momentjs.com/docs/#/displaying/format/ for formatting options.
export type DateFormat = string | false; // When false it should not be used
export type TimeFormat = string | false; // When false it should not be used
