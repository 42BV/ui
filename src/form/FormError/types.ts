export type OnChange = (hasErrors: boolean) => void;