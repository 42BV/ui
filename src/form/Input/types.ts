export type Position = 'left' | 'right';
export type Mask = (string | RegExp)[];
