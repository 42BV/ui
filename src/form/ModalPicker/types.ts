export type AddButtonCallback<T> = () => Promise<T>;

export interface AddButtonOptions<T> {
  label: string;
  onClick: AddButtonCallback<T>;
}
