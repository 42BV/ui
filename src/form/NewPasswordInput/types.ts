export type NewPasswordInputRule =
  | 'lowercase'
  | 'uppercase'
  | 'number'
  | 'specialChar'
  | 'minimumLength'
  | 'noSpace';
