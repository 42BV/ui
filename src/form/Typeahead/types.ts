/**
 * Represents the value of a Typeahead.
 */
export type TypeaheadOption<T> = {
  label: string;
  value: T;
};
