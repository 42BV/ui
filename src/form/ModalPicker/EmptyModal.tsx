import React from 'react';
import { ContentState } from '../../core/ContentState/ContentState';
import { t } from '../../utilities/translation/translation';

type Props = {
  /**
   * Whether or not the user performed a search.
   */
  userHasSearched: boolean;
  title?: string;
  subTitle?: string;
};

/**
 * The EmptyModal is a component which is rendered when a
 * Modal is empty. Can either display a message saying that
 * it is empty when there is nothing to be selected. Or a
 * message saying that there is no result for the current
 * search query when the user has searched.
 */
export function EmptyModal(props: Props) {
  const { userHasSearched, title, subTitle } = props;
  if (userHasSearched) {
    return (
      <ContentState
        mode="no-results"
        title={t({
          key: 'EmptyModal.NO_RESULTS.TITLE',
          fallback: 'Empty',
          overrideText: title
        })}
        subTitle={t({
          key: 'EmptyModal.NO_RESULTS.SUBTITLE',
          fallback:
            'No results were found please try again with a different query.',
          overrideText: subTitle
        })}
      />
    );
  } else {
    return (
      <ContentState
        mode="empty"
        title={t({
          key: 'EmptyModal.EMPTY.TITLE',
          fallback: 'Empty',
          overrideText: title
        })}
        subTitle={t({
          key: 'EmptyModal.EMPTY.SUBTITLE',
          fallback: 'There is nothing here yet, the collection is empty.',
          overrideText: subTitle
        })}
      />
    );
  }
}
