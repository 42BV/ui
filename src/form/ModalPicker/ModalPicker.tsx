import React from 'react';
import { Page } from '@42.nl/spring-connect';
import {
  Modal,
  ModalHeader,
  ModalBody,
  ModalFooter,
  Button,
  Row,
  Col
} from 'reactstrap';

import Pagination from '../../core/Pagination/Pagination';
import { t } from '../../utilities/translation/translation';
import SearchInput from '../../core/SearchInput/SearchInput';

interface Text {
  placeholder?: string;
  cancel?: string;
  select?: string;
}

interface Props {
  /**
   * The placeholder of the form element.
   */
  placeholder: string;

  /**
   * Whether or not the modal is open.
   */
  isOpen: boolean;

  /**
   * The current page of options. Used to determine if the
   * pagination should be rendered.
   */
  page: Page<any>;

  /**
   * Here the component using the ModalPicker must render in the options.
   */
  children: React.ReactNode;

  /**
   * The value to show in the search input.
   */
  query: string;

  /**
   * Whether or not to show the search input.
   */
  canSearch: boolean;

  /**
   * Callback to fetch the options to display to the user.
   */
  fetchOptions: (query: string) => void;

  /**
   * Callback for when the page changes.
   */
  pageChanged: (page: number) => void;

  /**
   * Callback for when the modal should close.
   */
  closeModal: () => void;

  /**
   * Callback for when the modal is saved
   */
  modalSaved: () => void;

  /**
   * Whether or not the save button is enabled
   */
  saveButtonEnabled: boolean;

  /**
   * Optionally an add button to display in the Modal. Can
   * be used to dynamically add an option which was not there
   * before.
   */
  addButton?: {
    label: string;
    onClick: () => void;
  };

  /**
   * Optionally customized text within the component.
   * This text should already be translated.
   */
  text?: Text;
}

/**
 * The ModalPicker component is an abstraction used by the ModalPickerMultiple
 * and ModalPickerSingle to render a modal. It contains the code which
 * could be shared by the two components.
 */
export default function ModalPicker(props: Props) {
  const {
    placeholder,
    isOpen,
    page,
    children,
    query,
    canSearch,
    fetchOptions,
    pageChanged,
    closeModal,
    modalSaved,
    saveButtonEnabled,
    addButton,
    text = {}
  } = props;

  const shouldRenderPagination = !(page.first && page.last);

  return (
    <Modal isOpen={isOpen} toggle={() => closeModal()}>
      <ModalHeader toggle={() => closeModal()}>{placeholder}</ModalHeader>
      <ModalBody>
        {canSearch ? (
          <Row>
            <Col>
              <SearchInput
                value={query}
                placeholder={t({
                  overrideText: text.placeholder,
                  key: 'ModalPicker.SEARCH',
                  fallback: 'Search...'
                })}
                onChange={value => fetchOptions(value)}
              />
            </Col>
          </Row>
        ) : null}
        {children}
      </ModalBody>

      {shouldRenderPagination ? (
        <ModalFooter className="d-flex justify-content-center">
          <Pagination page={page} onChange={page => pageChanged(page)} />
        </ModalFooter>
      ) : null}

      <ModalFooter className="d-flex flex-row justify-content-between">
        {addButton ? (
          <Button color="primary" onClick={() => addButton.onClick()}>
            {addButton.label}
          </Button>
        ) : (
          <div />
        )}

        <div>
          <Button
            className="ml-1"
            color="secondary"
            onClick={() => closeModal()}
          >
            {t({
              overrideText: text.cancel,
              key: 'ModalPicker.CANCEL',
              fallback: 'Cancel'
            })}
          </Button>
          <Button
            className="ml-1"
            color="primary"
            onClick={() => modalSaved()}
            disabled={!saveButtonEnabled}
          >
            {t({
              overrideText: text.select,
              key: 'ModalPicker.SELECT',
              fallback: 'Select'
            })}
          </Button>
        </div>
      </ModalFooter>
    </Modal>
  );
}
