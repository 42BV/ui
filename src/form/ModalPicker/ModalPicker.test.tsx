import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { emptyPage, Page } from '@42.nl/spring-connect';

import ModalPicker, { RenderOptionsConfig } from './ModalPicker';
import { User } from '../../test/types';
import { adminUser, coordinatorUser, pageOfUsers, userUser } from '../../test/fixtures';
import { ModalPickerRenderOptionsOption } from './types';
import { ListGroup, ListGroupItem } from 'reactstrap';
import lodash from 'lodash';

describe('Component: ModalPicker', () => {
  function setup({
    showAddButton,
    canSearch = true,
    page = pageOfUsers(),
    loading = false,
    renderOptionsConfig,
    canSearchSync = true
  }: {
    page?: Page<User>;
    showAddButton?: boolean;
    canSearch?: boolean;
    loading?: boolean;
    renderOptionsConfig?: RenderOptionsConfig<User>;
    canSearchSync?: boolean;
  }) {
    const queryChangedSpy = jest.fn();
    const pageChangedSpy = jest.fn();
    const closeModalSpy = jest.fn();
    const modalSavedSpy = jest.fn();
    const addButtonSpy = jest.fn();

    const addButton = showAddButton
      ? { label: 'Add color', onClick: addButtonSpy }
      : undefined;

    const { container } = render(
      <ModalPicker
        placeholder="Please select your favorite color"
        isOpen={true}
        page={page}
        queryChanged={queryChangedSpy}
        pageChanged={pageChangedSpy}
        closeModal={closeModalSpy}
        modalSaved={modalSavedSpy}
        selected={adminUser()}
        query=""
        canSearch={canSearch}
        addButton={addButton}
        userHasSearched={true}
        canSearchSync={canSearchSync}
        loading={loading}
        renderOptionsConfig={renderOptionsConfig}
      >
        <h1>Children</h1>
      </ModalPicker>
    );

    return {
      container,
      queryChangedSpy,
      pageChangedSpy,
      closeModalSpy,
      modalSavedSpy,
      addButtonSpy
    };
  }

  describe('ui', () => {
    test('default', () => {
      setup({});
      expect(document.body.lastChild).toMatchSnapshot();
    });

    test('without addButton and search', () => {
      setup({ canSearch: false, showAddButton: false });
      expect(screen.queryByText('Add color')).not.toBeInTheDocument();
      expect(screen.queryByRole('searchbox')).not.toBeInTheDocument();
    });

    test('with addButton and search', () => {
      setup({ canSearch: true, showAddButton: true });
      expect(screen.queryByText('Add color')).toBeInTheDocument();
      expect(screen.queryByRole('searchbox')).toBeInTheDocument();
    });

    test('without Pagination', () => {
      setup({ page: emptyPage() });
      expect(screen.queryByText('arrow_forward')).not.toBeInTheDocument();
    });

    test('with Pagination', () => {
      setup({});
      expect(screen.queryByText('arrow_forward')).toBeInTheDocument();
    });

    test('loading', () => {
      setup({ loading: true });
      expect(document.body.lastChild).toMatchSnapshot();
    });

    test('empty', () => {
      setup({ page: emptyPage() });
      expect(screen.queryByText('Empty')).toBeInTheDocument();
    });

    test('with children', () => {
      setup({});
      expect(screen.queryByText('Children')).toBeInTheDocument();
    });
  });

  describe('events', () => {
    it('should call closeModal when the user clicks the X in the modal header', () => {
      const { closeModalSpy } = setup({});

      fireEvent.click(screen.getByLabelText('Close'));

      expect(closeModalSpy).toHaveBeenCalledTimes(1);
    });

    it('should call closeModal when the user clicks cancel', () => {
      const { closeModalSpy } = setup({});

      fireEvent.click(screen.getByText('Cancel'));

      expect(closeModalSpy).toHaveBeenCalledTimes(1);
    });

    it('should call modalSaved when the user clicks select', () => {
      const { modalSavedSpy } = setup({});

      fireEvent.click(screen.getByText('Select'));

      expect(modalSavedSpy).toHaveBeenCalledTimes(1);
    });

    it('should call search when user stops typing', () => {
      // @ts-expect-error Test mock
      jest.spyOn(lodash, 'debounce').mockImplementation((fn) => {
        return fn;
      });

      const { queryChangedSpy } = setup({});

      fireEvent.change(screen.getByRole('searchbox'), { target: { value: 'Maarten' } });

      expect(queryChangedSpy).toHaveBeenCalledTimes(1);
      expect(queryChangedSpy).toHaveBeenCalledWith('Maarten');
    });

    it('should call pageChanged when the pagination is used', () => {
      const { pageChangedSpy } = setup({});

      fireEvent.click(screen.getByText('arrow_forward'));

      expect(pageChangedSpy).toHaveBeenCalledTimes(1);
      expect(pageChangedSpy).toHaveBeenCalledWith(3);
    });

    it('should call addButton when the add button is clicked', () => {
      const { addButtonSpy } = setup({
        showAddButton: true
      });

      fireEvent.click(screen.getByText('Add color'));

      expect(addButtonSpy).toHaveBeenCalledTimes(1);
      expect(addButtonSpy).toHaveBeenCalledWith();
    });
  });

  it('should render via renderOption and provide a working api for the developer when a renderOptionsConfig is provided', () => {
    const onChangeSpy = jest.fn();

    function renderOptions(
      options: ModalPickerRenderOptionsOption<User>[]
    ): React.ReactNode {
      return (
        <ListGroup>
          {options.map(
            ({
              option,
              toggle
            }: ModalPickerRenderOptionsOption<User>) => (
              <ListGroupItem
                key={option.id}
                onClick={toggle}
              >
                {option.email}
              </ListGroupItem>
            )
          )}
        </ListGroup>
      );
    }

    setup({
      renderOptionsConfig: {
        labelForOption: (user) => user.email,
        isOptionEnabled: (user) => user.email !== 'admin@42.nl',
        renderOptions,
        onChange: onChangeSpy
      }
    });

    expect(screen.getByText('admin@42.nl')).toHaveClass('list-group-item');

    fireEvent.click(screen.getByText('admin@42.nl'));
    expect(onChangeSpy).toBeCalledTimes(0);

    fireEvent.click(screen.getByText('coordinator@42.nl'));
    expect(onChangeSpy).toBeCalledTimes(1);
    expect(onChangeSpy).toBeCalledWith(coordinatorUser(), false);

    fireEvent.click(screen.getByText('user@42.nl'));
    expect(onChangeSpy).toBeCalledTimes(2);
    expect(onChangeSpy).toBeCalledWith(userUser(), false);
  });
});
