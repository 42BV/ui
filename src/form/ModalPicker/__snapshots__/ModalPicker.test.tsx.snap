// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: ModalPicker ui with addButton and search: Component: ModalPicker => ui with addButton and search 1`] = `
<Modal
  autoFocus={true}
  backdrop={true}
  backdropTransition={
    Object {
      "mountOnEnter": true,
      "timeout": 150,
    }
  }
  centered={false}
  container="body"
  fade={true}
  isOpen={true}
  keyboard={true}
  modalTransition={
    Object {
      "timeout": 300,
    }
  }
  onClosed={[Function]}
  onOpened={[Function]}
  returnFocusAfterClose={true}
  role="dialog"
  scrollable={false}
  toggle={[Function]}
  trapFocus={false}
  unmountOnClose={true}
  zIndex={1050}
>
  <ModalHeader
    closeAriaLabel="Close"
    tag="h5"
    toggle={[Function]}
    wrapTag="div"
  >
    Please select your favorite color
  </ModalHeader>
  <ModalBody
    tag="div"
  >
    <Row
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
          "xxl",
        ]
      }
    >
      <Col
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
            "xxl",
          ]
        }
      >
        <SearchInput
          debounce={0}
          defaultValue=""
          onChange={[MockFunction]}
          placeholder="Search..."
        />
      </Col>
    </Row>
    <Row
      className="mt-3"
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
          "xxl",
        ]
      }
    >
      <Col
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
            "xxl",
          ]
        }
      >
        <h1>
          Children
        </h1>
      </Col>
    </Row>
  </ModalBody>
  <ModalFooter
    className="d-flex justify-content-center"
    tag="div"
  >
    <Pagination
      onChange={[Function]}
      page={
        Object {
          "content": Array [
            Object {
              "active": true,
              "email": "admin@42.nl",
              "firstName": "Addie",
              "id": 42,
              "lastName": "Admin",
              "roles": Array [
                "ADMIN",
              ],
            },
            Object {
              "active": false,
              "email": "coordinator@42.nl",
              "firstName": "Cordelia",
              "id": 777,
              "lastName": "Coordinator",
              "roles": Array [
                "ADMIN",
                "USER",
              ],
            },
            Object {
              "active": false,
              "email": "user@42.nl",
              "firstName": "Ulysses",
              "id": 1337,
              "lastName": "User",
              "roles": Array [
                "USER",
              ],
            },
          ],
          "first": false,
          "last": false,
          "number": 2,
          "numberOfElements": 3,
          "size": 3,
          "totalElements": 9,
          "totalPages": 3,
        }
      }
    />
  </ModalFooter>
  <ModalFooter
    className="d-flex flex-row justify-content-between"
    tag="div"
  >
    <Button
      color="primary"
      onClick={[Function]}
      tag="button"
    >
      Add color
    </Button>
    <div>
      <Button
        className="ms-1"
        color="secondary"
        onClick={[Function]}
        tag="button"
      >
        Cancel
      </Button>
      <Button
        className="ms-1"
        color="primary"
        disabled={false}
        onClick={[Function]}
        tag="button"
      >
        Select
      </Button>
    </div>
  </ModalFooter>
</Modal>
`;

exports[`Component: ModalPicker ui without addButton and search: Component: ModalPicker => ui without addButton and search 1`] = `
<Modal
  autoFocus={true}
  backdrop={true}
  backdropTransition={
    Object {
      "mountOnEnter": true,
      "timeout": 150,
    }
  }
  centered={false}
  container="body"
  fade={true}
  isOpen={true}
  keyboard={true}
  modalTransition={
    Object {
      "timeout": 300,
    }
  }
  onClosed={[Function]}
  onOpened={[Function]}
  returnFocusAfterClose={true}
  role="dialog"
  scrollable={false}
  toggle={[Function]}
  trapFocus={false}
  unmountOnClose={true}
  zIndex={1050}
>
  <ModalHeader
    closeAriaLabel="Close"
    tag="h5"
    toggle={[Function]}
    wrapTag="div"
  >
    Please select your favorite color
  </ModalHeader>
  <ModalBody
    tag="div"
  >
    <Row
      className="mt-3"
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
          "xxl",
        ]
      }
    >
      <Col
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
            "xxl",
          ]
        }
      >
        <h1>
          Children
        </h1>
      </Col>
    </Row>
  </ModalBody>
  <ModalFooter
    className="d-flex justify-content-center"
    tag="div"
  >
    <Pagination
      onChange={[Function]}
      page={
        Object {
          "content": Array [
            Object {
              "active": true,
              "email": "admin@42.nl",
              "firstName": "Addie",
              "id": 42,
              "lastName": "Admin",
              "roles": Array [
                "ADMIN",
              ],
            },
            Object {
              "active": false,
              "email": "coordinator@42.nl",
              "firstName": "Cordelia",
              "id": 777,
              "lastName": "Coordinator",
              "roles": Array [
                "ADMIN",
                "USER",
              ],
            },
            Object {
              "active": false,
              "email": "user@42.nl",
              "firstName": "Ulysses",
              "id": 1337,
              "lastName": "User",
              "roles": Array [
                "USER",
              ],
            },
          ],
          "first": false,
          "last": false,
          "number": 2,
          "numberOfElements": 3,
          "size": 3,
          "totalElements": 9,
          "totalPages": 3,
        }
      }
    />
  </ModalFooter>
  <ModalFooter
    className="d-flex flex-row justify-content-between"
    tag="div"
  >
    <div />
    <div>
      <Button
        className="ms-1"
        color="secondary"
        onClick={[Function]}
        tag="button"
      >
        Cancel
      </Button>
      <Button
        className="ms-1"
        color="primary"
        disabled={false}
        onClick={[Function]}
        tag="button"
      >
        Select
      </Button>
    </div>
  </ModalFooter>
</Modal>
`;
